### PR TITLE
[jit] Add some of nn.init to weak script

### PR DESCRIPTION
**Pull request by driazati**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#19640 [jit] Add some of nn.init to weak script**

Differential Revision: [D15065332](https://our.internmc.facebook.com/intern/diff/15065332/)

### CODE REVIEW

File: torch/nn/init.py [hunk 1]
```diff
@@ -105,10 +135,11 @@ def ones_(tensor):
         >>> w = torch.empty(3, 5)
         >>> nn.init.ones_(w)
     """
-    with torch.no_grad():
-        return tensor.fill_(1)
+    # type: (Tensor) -> Tensor
+    # return _no_grad_fill_(tensor, 1.)
```
Comment (eellison) [inline torch/nn/init.py]: Accidental comment out

File: torch/csrc/jit/register_special_ops.cpp [hunk 2]
```diff
@@ -395,7 +395,48 @@ RegisterOperators reg(
...
+       torch::NoGradGuard no_grad;
+
+       at::Tensor tensor;
+       double mean;
+       double std;
+       pop(stack, tensor, mean, std);
+       push(stack, at::_th_normal_(tensor, mean, std));
+       return 0;
+     }),
+     Operator("aten::_no_grad_fill_(Tensor(a) tensor, float val) -> Tensor(a)", [](Stack& stack) {
```
Comment (eellison) [inline torch/csrc/jit/register_special_ops.cpp]: It should be Tensor(a!) since it is mutated

File: torch/csrc/jit/register_special_ops.cpp [hunk 3]
```diff
@@ -395,7 +395,48 @@ RegisterOperators reg(
      Operator(
          "aten::_pack_sequence(Tensor output, Tensor batch_sizes, Tensor? sorted_indices, "
          "Tensor? unsorted_indices) -> (Tensor, Tensor, Tensor?, Tensor?)",
-         [](Stack& stack) { return 0; })
+         [](Stack& stack) { return 0; }),
+     Operator("aten::_no_grad_uniform_(Tensor(a) tensor, float a, float b) -> Tensor(a)", [](Stack& stack) {
```
Comment (eellison) [inline torch/csrc/jit/register_special_ops.cpp]: This should be mutated as well right ?

File: torch/csrc/jit/register_special_ops.cpp [hunk 4]
```diff
@@ -395,7 +395,48 @@ RegisterOperators reg(
...
+       // TODO: remove when script supports setting grad mode
+       torch::NoGradGuard no_grad;
+
+       at::Tensor tensor;
+       double val;
+       pop(stack, tensor, val);
+       push(stack, at::fill_(tensor, val));
+       return 0;
+     }),
+     Operator("aten::_no_grad_zero_(Tensor(a) tensor) -> Tensor(a)", [](Stack& stack) {
```
Comment (eellison) [inline torch/csrc/jit/register_special_ops.cpp]: mutated

File: torch/nn/init.py [hunk 5]
```diff
@@ -241,10 +276,11 @@ def xavier_normal_(tensor, gain=1):
         >>> w = torch.empty(3, 5)
         >>> nn.init.xavier_normal_(w)
     """
+    # type: (Tensor, float) -> Tensor
     fan_in, fan_out = _calculate_fan_in_and_fan_out(tensor)
-    std = gain * math.sqrt(2.0 / (fan_in + fan_out))
-    with torch.no_grad():
-        return tensor.normal_(0, std)
+    std = gain * math.sqrt(2.0 / float(fan_in + fan_out))
```
Comment (eellison) [inline torch/nn/init.py]: Why did you need the float cast ?

File: test/test_jit.py [hunk 6]
```diff
@@ -11193,6 +11193,34 @@ def nest_for_ret(x):
...
+                def test({}):
+                    # type: ({})
+                    return torch.nn.init.{}({})
+            ''').format(arg_str, type_str, name, arg_str)
+            cu = torch.jit.CompilationUnit(code)
+
+            # Compare functions
+            init_fn = getattr(torch.nn.init, name)
+            script_out = self.runAndSaveRNG(cu.test, args_fn())
+            eager_out = self.runAndSaveRNG(init_fn, args_fn())
```
Comment (eellison) [inline test/test_jit.py]: Can you add an assertion that there is no python op

File: test/test_jit.py [hunk 7]
```diff
@@ -11193,6 +11193,34 @@ def nest_for_ret(x):
...
+            ('xavier_normal_', (lambda: (torch.ones(2, 2),)), "Tensor"),
+            ('xavier_uniform_', (lambda: (torch.ones(2, 2),)), "Tensor"),
+        )
+
+        for name, args_fn, type_str in tests:
+            # Build test code
+            arg_str = ', '.join([chr(i + ord('a')) for i in range(len(args_fn()))])
+
+            code = dedent('''
+                def test({}):
```
Comment (eellison) [inline test/test_jit.py]: nit: a little more readable if you name the formatting slots

File: torch/nn/init.py (hunk 5 above)
Comment (driazati) [reply torch/nn/init.py]: `fan_in` and `fan_out` are ints, and there's no `aten::div(float, int)` currently

Comment (eellison) [reply torch/nn/init.py]: Can you add that to this PR? it should be like a two line change - `DEFINE_INT_FLOAT_OP(aten::div, a / b, float)` in register_prim_ops and add '/' to test_number_math.

Comment (eellison) [reply torch/nn/init.py]: This wasn't really resolved.

File: test/test_jit.py [hunk 8]
```diff
@@ -11193,6 +11193,36 @@ def nest_for_ret(x):
                         return 4
                 return 5
 
+    def test_nn_init(self):
+        tests = (
+            ('constant_', (lambda: (torch.ones(2, 2), 2.5)), "Tensor, float"),
```
Comment (eellison) [inline test/test_jit.py]: Can you switch it to rand ?

**Review by eellison — APPROVED** [review]
LGTM. A little more coverage on different inputs would be nice, as would be adding the div float / int, which should be a really really easy change.

File: torch/nn/init.py (hunk 5 above)
Comment (driazati) [reply torch/nn/init.py]: At least should be left for a follow up, I feel like it counts as an implicit `int` -> `float` conversion which if we're going to add should be at the compiler level instead of per-op.

File: torch/nn/init.py [hunk 9]
```diff
@@ -105,11 +136,12 @@ def ones_(tensor):
         >>> w = torch.empty(3, 5)
         >>> nn.init.ones_(w)
     """
-    with torch.no_grad():
-        return tensor.fill_(1)
+    return _no_grad_fill_(tensor, 1.)
```
Comment (eellison) [inline torch/nn/init.py]: Hmm it looks like this is changing the semantics, because previously the type of the tensor would have been preserved, but now it is being filled with floats
